hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderSelectBuilder.java

```diff
@@ -419,15 +426,21 @@ private void applyOrdering(TableGroup tableGroup, OrderByFragment orderByFragmen
 		log.tracef( "Starting visitation of FetchParent's Fetchables : %s", fetchParent.getNavigablePath() );
 
 		final List<Fetch> fetches = new ArrayList<>();
+		String fullPath = fetchParent.getNavigablePath().getFullPath();
+		if ( visitedNavigablePath.containsKey( fullPath ) ) {
+			return visitedNavigablePath.get( fullPath );
+		}
 
 		final BiConsumer<Fetchable, Boolean> processor = createFetchableBiConsumer( fetchParent, querySpec, creationState, fetches );
```

Comment:
moving the declaration of 'fetches' before the above statement would improve code readability a little bit. As a general rule, putting variable declaration as near its usage as possible decrease human brain memorizing pressure, I think.